soroban-env-host/src/host/conversion.rs

```diff
@@ -346,6 +342,23 @@ impl Host {
         })
     }
 
+    // Version of `to_host_val` for the internal cases where the value has to
+    // be valid by construction (e.g. read from ledger).
+    pub(crate) fn to_valid_host_val(&self, v: &ScVal) -> Result<Val, HostError> {
```

Comment:
I would have thought it'd just always be safer to do the check. Why do we need the optimization?

Reply:
Which check? This is not an optimization, just an 'assertion'. It's not absolutely necessary, but seems pretty useful. Alternative is to just do all the conversions via `to_host_val`, which has a risk of not noticing a real internal error.